cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/AbstractStateChangeProgramController.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.program;
+
+import co.cask.cdap.app.runtime.ProgramController;
+import co.cask.cdap.app.runtime.ProgramStateWriter;
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.internal.app.runtime.AbstractListener;
+import co.cask.cdap.internal.app.runtime.AbstractProgramController;
+import co.cask.cdap.proto.ProgramRunStatus;
+import co.cask.cdap.proto.id.ProgramRunId;
+import com.google.common.util.concurrent.Service;
+import org.apache.twill.common.Threads;
+import org.apache.twill.internal.ServiceListenerAdapter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+/**
+ * Abstract implementation of {@link AbstractProgramController} that responds to state transitions and persists all
+ * state changes.
+ */
+public abstract class AbstractStateChangeProgramController extends AbstractProgramController {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractStateChangeProgramController.class);
+
+  public AbstractStateChangeProgramController(Service service, final ProgramRunId programRunId, final String twillRunId,
+                                              final ProgramStateWriter programStateWriter,
+                                              @Nullable String componentName) {
+    super(programRunId.getParent(), RunIds.fromString(programRunId.getRun()), componentName);
+
+    service.addListener(
```

Comment:
I'm not entirely sure. `DynamicPartitioningTestRun` fails for example, with issues seemingly unrelated with dataset tables not being found.

Removing this constructor would "break" / change behavior from what's currently happening. For Spark / MR, we were adding the listener on the service.